crates/wasmtime/src/runtime/gc/enabled/externref.rs

```diff
@@ -0,0 +1,715 @@
+/// ```
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct ExternRef {
+    inner: GcRootIndex,
+}
+
+unsafe impl GcRefImpl for ExternRef {
+    #[allow(private_interfaces)]
+    fn transmute_ref(index: &GcRootIndex) -> &Self {
+        // Safety: `ExternRef` is a newtype of a `GcRootIndex`.
+        let me: &Self = unsafe { std::mem::transmute(index) };
+
+        // Assert we really are just a newtype of a `GcRootIndex`.
+        assert!(matches!(
+            me,
+            Self {
+                inner: GcRootIndex { .. },
+            }
+        ));
+
+        me
+    }
+}
+
+impl ExternRef {
+    /// Creates a new instance of `ExternRef` wrapping the given value.
+    ///
+    /// The resulting value is automatically unrooted when the given `context`'s
+    /// scope is exited. See [`Rooted<T>`][crate::Rooted]'s documentation for
+    /// more details.
+    ///
+    /// This operation may trigger a GC.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use wasmtime::*;
+    /// # fn _foo() -> Result<()> {
+    /// let mut store = Store::<()>::default();
+    ///
+    /// {
+    ///     let mut scope = RootScope::new(&mut store);
+    ///
+    ///     // Create an `externref` wrapping a `str`.
+    ///     let externref = ExternRef::new(&mut scope, "hello!");
+    ///
+    ///     // Use `externref`...
+    /// }
+    ///
+    /// // The `externref` is automatically unrooted when we exit the scope.
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn new<T>(mut context: impl AsContextMut, value: T) -> Result<Rooted<ExternRef>>
+    where
+        T: 'static + Any + Send + Sync,
+    {
+        let ctx = context.as_context_mut().0;
+
+        let value: Box<dyn Any + Send + Sync> = Box::new(value);
+
+        let gc_ref = match ctx
+            .gc_store_mut()
+            .alloc_externref(value)
+            .context("unrecoverable error when allocating new `externref`")?
+        {
+            Ok(gc_ref) => gc_ref,
+            Err(value) => {
+                // On recoverable allocation failure, do a GC and then try to
+                // allocate again.
+                ctx.gc();
+                ctx.gc_store_mut()
+                    .alloc_externref(value)
+                    .context("unrecoverable error when allocating new `externref`")?
+                    .map_err(|_| anyhow!("failed to allocate `externref`: GC heap out of memory"))?
+            }
+        };
+
+        let mut ctx = AutoAssertNoGc::new(ctx);
+        Ok(Self::from_cloned_gc_ref(&mut ctx, gc_ref.into()))
+    }
+
+    /// Construct a new `ExternRef` asynchronously.
+    ///
+    /// This is the same as `ExternRef::new` but if the allocation triggers a
+    /// GC, the GC is performed asynchronously, periodically yielding control to
+    /// the event loop.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given context was not configured to enable async support.
+    #[cfg(feature = "async")]
+    pub async fn new_async<S, T>(
```

Comment:
To avoid new/new_async, could the auto-gc here be pushed up to the caller, e.g. communicated through the `Result`?

Reply:
To clarify, you mean define a new error type like `GcHeapOutOfMemory` and return that, allowing callers to decide whether they want to do a GC and try again, rather than automatically doing a GC on their behalf?

Reply:
Indeed!

Reply:
Hm actually I'm not sure how this would work, because if the `externref` allocation fails, we need to give the host value back to the caller. We could stuff it back into the `GcHeapOutOfMemory` error, but then we would need a type parameter on the error type, and things start to get pretty funky. Not completely unworkable, but pretty funky. And then we need a different error type for non-`externref` errors that don't have a host value to return to the caller on OOM.

Do you still think it is worth pursuing this?

Reply:
Personally if you're ok with it I'd prefer to stick to the route of a custom error, basicaly `pub GcHeapOutOfMemory<T>(pub T)`. Otherwise I'd fear that there's little chance of embedders remembering to use `new_async` vs `new`